File: src/test/codewhisperer/util/bm25.test.ts
/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import assert from 'assert'
import { BM25Okapi } from '../../../codewhisperer/util/supplementalContext/rankBm25'

describe('bm25', function () {
    it('simple case 1', function () {
        const query = 'windy London'
        const corpus = ['Hello there good man!', 'It is quite windy in London', 'How is the weather today?']

        const sut = new BM25Okapi(corpus)
        const actual = sut.score(query)

        assert.deepStrictEqual(actual, [
            {
                content: 'Hello there good man!',
                index: 0,
                score: 0,
            },
            {
                content: 'It is quite windy in London',
                index: 1,
                score: 0.937294722506405,
            },
            {
                content: 'How is the weather today?',
                index: 2,
                score: 0,
            },
        ])

        assert.deepStrictEqual(sut.topN(query, 1), [
            {
                content: 'It is quite windy in London',
                index: 1,
                score: 0.937294722506405,
            },
        ])
    })

    it('simple case 2', function () {
        const query = 'codewhisperer is a machine learning powered code generator'
        const corpus = [
            'codewhisperer goes GA at April 2023',
            'machine learning tool is the trending topic!!! :)',
            'codewhisperer is good =))))',
            'codewhisperer vs. copilot, which code generator better?',
            'copilot is a AI code generator too',
            'it is so amazing!!',
        ]

        const sut = new BM25Okapi(corpus)
        const actual = sut.score(query)

        assert.deepStrictEqual(actual, [
            {
                content: 'codewhisperer goes GA at April 2023',
                index: 0,
                score: 0,
            },
            {
                content: 'machine learning tool is the trending topic!!! :)',
                index: 1,
                score: 2.597224531416621,
            },
            {
                content: 'codewhisperer is good =))))',
                index: 2,
                score: 0.3471790843435529,
            },
            {
                content: 'codewhisperer vs. copilot, which code generator better?',
                index: 3,
                score: 1.063018436525109,
            },
            {
                content: 'copilot is a AI code generator too',
                index: 4,
                score: 2.485359418462239,
            },
            {
                content: 'it is so amazing!!',
                index: 5,
                score: 0.3154033715392277,
            },
        ])

        assert.deepStrictEqual(sut.topN(query, 1), [
            {
                content: 'machine learning tool is the trending topic!!! :)',
                index: 1,
                score: 2.597224531416621,
            },
        ])

        assert.deepStrictEqual(sut.topN(query, 3), [
            {
                content: 'machine learning tool is the trending topic!!! :)',
                index: 1,
                score: 2.597224531416621,
            },
            {
                content: 'copilot is a AI code generator too',
                index: 4,
                score: 2.485359418462239,
            },
            {
                content: 'codewhisperer vs. copilot, which code generator better?',
                index: 3,
                score: 1.063018436525109,
            },
        ])
    })
})

File: plugins/amazonq/codewhisperer/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/codewhisperer/Bm25Test.kt
// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.jetbrains.services.codewhisperer

import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import software.aws.toolkits.jetbrains.services.codewhisperer.util.BM25
import software.aws.toolkits.jetbrains.services.codewhisperer.util.BM250kapi
import software.aws.toolkits.jetbrains.services.codewhisperer.util.BM25Result

class Bm25Test {
    private lateinit var sut: BM25
    private lateinit var tokenizer: (String) -> List<String>

    @Before
    fun setup() {
        tokenizer = { str: String ->
            str.split(" ")
        }
    }

    @Test
    fun `scores should return all scores and topN should return only highest n - simple case 1`() {
        val query = "windy London"
        val corpus = listOf(
            "Hello there good man!",
            "It is quite windy in London",
            "How is the weather today?"
        )

        sut = BM250kapi(corpus, tokenizer)
        val scores = sut.score(query)

        assertThat(scores).isEqualTo(
            listOf(
                BM25Result(docString = "Hello there good man!", score = 0.0),
                BM25Result(docString = "It is quite windy in London", score = 0.937294722506405),
                BM25Result(docString = "How is the weather today?", score = 0.0),
            )
        )

        assertThat(sut.topN(query, 1)).isEqualTo(
            listOf(
                BM25Result(docString = "It is quite windy in London", score = 0.937294722506405)
            )
        )
    }

    @Test
    fun `scores should return correct pair of document to its calculated score - simple case 2`() {
        val query = "codewhisperer is a machine learning powered code generator"
        val corpus = listOf(
            "codewhisperer goes GA at April 2023",
            "machine learning tool is the trending topic!!! :)",
            "codewhisperer is good =))))",
            "codewhisperer vs. copilot, which code generator better?",
            "copilot is a AI code generator too",
            "it is so amazing!!"
        )

        sut = BM250kapi(corpus, tokenizer)
        val scores = sut.score(query)

        assertThat(scores).isEqualTo(
            listOf(
                BM25Result(docString = "codewhisperer goes GA at April 2023", score = 0.0),
                BM25Result(docString = "machine learning tool is the trending topic!!! :)", score = 2.5075225906025422),
                BM25Result(docString = "codewhisperer is good =))))", score = 0.33539413050870837),
                BM25Result(docString = "codewhisperer vs. copilot, which code generator better?", score = 1.0935565858644076),
                BM25Result(docString = "copilot is a AI code generator too", score = 2.5673872789459544),
                BM25Result(docString = "it is so amazing!!", score = 0.33539413050870837)
            )
        )

        assertThat(sut.topN(query, 1)).isEqualTo(
            listOf(
                BM25Result(docString = "copilot is a AI code generator too", score = 2.5673872789459544)
            )
        )
        assertThat(sut.topN(query, 3)).isEqualTo(
            listOf(
                BM25Result(docString = "copilot is a AI code generator too", score = 2.5673872789459544),
                BM25Result(docString = "machine learning tool is the trending topic!!! :)", score = 2.5075225906025422),
                BM25Result(docString = "codewhisperer vs. copilot, which code generator better?", score = 1.0935565858644076),
            )
        )
    }
}
